### DISCUSSION TITLE
JWT token validation: Centralized vs Decentralized

### DISCUSSION THREAD

**Discussion by withinoneyear** [zitadel/zitadel#8100] [Q&A]
According to this [document](https://zitadel.com/docs/guides/integrate/service-users/authenticate-service-users), `private key JWT authentication` is recommended choice, and it also lists the benefits, especially it states that:

> Decentralized Verification: No need for constant server calls, improving performance and scalability.

But in the [document](https://zitadel.com/docs/guides/integrate/token-introspection/private-key-jwt#token-introspection) of token introspection, it requires the resource server to call Zitadel endpoint to validate the token.

So is it required to always use Zitadel to validate the token, or is it just an option? Can jwks be used to locally verify the token instead? If it is required, it makes the entire process centralized, not only it impacts the performance, but also heavily depends on the Zitadel performance and scalalbility.

**Comment by hifabienne** [comment]
@livio-a @muhlemmer can you help here?

**Comment by muhlemmer** [comment]
@withinoneyear JWT tokens can indeed be fully verified by the client after it gets the public keys from authorization server. Keys should be cached by the client, and the client should re-get keys once it gets an token where the `kid` is not part of the current cached keys. This way you can do semi-decentralized verification of tokens.

You can choose to call the introspection endpoint, but this is not a requirement. When the client does not use introspection, terminated sessions or revoked tokens will not be noticed by the client. It's up to you if the client should care. We recommend caching introspection responses for a time that is acceptable for the application and session termination does not need to be instant. This is a hybrid solution between not checking at all or on every request.

**Reply by danielloader** [reply]
I wish Istio implemented this, they don't invalidate the cache when the `kid` mismatches (implying it's worth trying to see if there's a new key). 

It's causing me a lot of sadness right now.

**Reply by fforootd** [reply]
Yeah we heard that in the past already and this also influenced the discussion about key rotation here https://github.com/zitadel/zitadel/discussions/7464
